amperser/proselint
# Dev-Merge [WIP]

Latest changes:

- remove memoize-wrapper as it adds a level of complexity to a user-interface
- lint() now only accepts string as input, lint_path() handles the previous options
- fix broken offset-magic done in check_functions
- further improve usability of result-processing functions
- major speedup especially multicpu in windows x3.3, serial x2
- new inside out regex-ing for less computational overhead
  - example orig: `(?:^|\W)w1[\W$]|(?:^|\W)w2[\W$]|(?:^|\W)w3[\W$]`
  - example new:  `(?:^|\W)(w1|w2|w3)[\W$]`

```shell
> lint_path(corpora)   OLD v0.15
serial took 50196.394 ms -> run0
serial took 53044.826 ms -> run1
parallel took 15721.931 ms -> run0
parallel took 15773.215 ms -> run1

> lint_path(corpora)  	optimized regex (NEW)
serial took 31550.519 ms -> run0
serial took 36539.587 ms -> run1
parallel took 10411.408 ms -> run0
parallel took 10132.006 ms -> run1
```

- add benchmark to compare lint(), lint_path() and proselint-executable
- fix and improve naming of checks (files, dirs and inside checks)
- add pre-commit tests-suite
- extend and activate check for roman numerals
- extend check for lexical illusions (now 1-4 words)
- add check for lower case word at the beginning of new sentence
- add check for comma between digits
- add check for repeated beginning word of consecutive sentences (monotonic)
- add check for no whitespace between numbers and its unit (scientific)
- add check for whitespace before punctuation
- add scientific-checks, ported from TNT https://github.com/entorb/typonuketool and other sources
- add check for unmatched braces (), [], {}
- add check for consistence of apostrophes
- fix checks for punctuation
- add check for digits at the beginning of sentence, and single digits
- improve various checks - mostly reduction of computational overhead of regexes
- move example text into check-files, these get tested by pytest
	- this approach is unconventional, but it's better maintainable (there were several tests missing, some were broken or testing same thing twice)
	- also better documentation
	- downside: slightly slower import of check-modules (unconfirmed)
- precompiling the regex of the slowest check improves performance significantly
- fix and improve several checks (less computation, more powerful, pass unittests)
	- lexical illusions
	- but
	- cursing (something like "a.s.s." allowed "amsisa")
- speedup for not needed safe_join-formatting (+10% in sherlock)
- speedup for sep_in_txt by setting word-boundry (\b) (+33% in sherlock)
- speedup by replacing sep_in_txt completely (100x less computation in sherlock)
- sherlock-benchmark of checks from >500s down to 174s
- reduce false positives on numbers-checks, use non-breaking space if you need to use numbers

latest benchmark-results, comparable to results above

```
############# lint_path(corpora) - Linux, 147 checks
serial took 26687.769 ms -> run0
serial took 28858.442 ms -> run1
parallel took 9076.079 ms -> run0
parallel took 8898.538 ms -> run1

```
